### PR TITLE
Add overrides for LoadMany to support List<T> as opposed to T[].

### LINKED ISSUE #1484: Allow passing `IEnumerable<T>` to `IDocumentSession.Store()`
This applies to the both the untennanted and tennanted methods.

Currently they only accept `params T[]`, there should be versions which take `IEnumerable<T>`

### PR DESCRIPTION
**Pull request by apereiratl**
Fixes #1484

### CODE REVIEW

**Comment by jeremydmiller** [comment]
I vote extension methods as well. I'm going to say again that I'm really not enthusiastic about these going into our core.

**Comment by oskardudycz** [comment]
@jeremydmiller I agree, that this is something that probably our user could write on their own (eg. by adding the extensions methods on their codebase), but if that helps the migration for the RavenDB users (that I assume were our initial userbase) then I think that it's fine. I think that if it's in the different namespace than the original interface then it shouldn't be a huge maintenance burden.

**Comment by mysticmind** [comment]
I echo the same thought as @jeremydmiller  and @oskardudycz . We could possibly look at adding a docs section to outline these so that users porting code base from RavenDB to Marten.

@jeremydmiller @oskardudycz should we close this PR or wait to hear from @apereiratl?

**Comment by oskardudycz** [comment]
I'd propose to wait a few days and see if @apereiratl still would like to come up with the extension methods solution.

**Comment by apereiratl** [comment]
Sorry I've been a bit busy but will convert them to extension methods and resend the patch.

**Comment by oskardudycz** [comment]
@apereiratl there is no rush on that. We just wanted to know if we may expect the updates in the nearest future 👍

**Comment by mysticmind** [comment]
@apereiratl we are planning to release v3.12.0 on 8th, is it something which you could look to squeeze in?

**Comment by oskardudycz** [comment]
@jeremydmiller @mysticmind I checked and unfortunately, it won't work with extension methods. Params methods will still take precedence and IEnumerable or IList will be used.

I think that it's fine to add such methods, as it's a common error of wrongly passing IEnumerable - especially in Append events function.

Same case for https://github.com/JasperFx/marten/pull/1485 provided by @trullock.

After consideration, I'm fine with that (however our interface will be even bigger). Probably suggestion made by @trullock to make the IEnumerable default method and params extension method would also work, but such a switch can be made in v4.

I can add necessary changes.

Thoughts?

**Comment by jeremydmiller** [comment]
I don't think this adds a huge amount of value to worry about what's an extension method and what's not. I say just add the overloads to `LoadMany()` and call it a day. 

I do want to reevaluate `IDocumentSession` in v4, but there I'm more worried about moving the plain text searching out of the root interface.

**Comment by oskardudycz** [comment]
I was also referring to other places where we have params [] eg. Store, Append events etc. 
I think that having a smaller interface and implementation and overloads as Extension methods might make it easier to maintain and less prone for breaking changes.

So if there are no votes against I’ll da overloads and we’ll evaluate removing some of them in V4

**Comment by jeremydmiller** [comment]
@oskardudycz Gotcha, so just that. It's a breaking binary change, so you have to leave the changes to thin it down for v4

**Comment by oskardudycz** [comment]
@jeremydmiller sounds good to nr 👍

**Comment by oskardudycz** [comment]
@jeremydmiller @mysticmind @jokokko I cherrypicked @trullock changes, aligned both implementations and added other overloads eg. to Insert, Update and added for EventStore.

Could you verify if you're fine with that?

File: src/Marten/QuerySession.cs [hunk 1]
```diff
@@ -119,13 +119,10 @@ private T load<T>(object id)
...
             var mapping = Tenant.MappingFor(typeof(T));
-            if (id.GetType() != mapping.IdType)
+            if (id.GetType() != mapping.IdType && (id.GetType() != typeof(int) || mapping.IdType != typeof(long)))
             {
-                if (id.GetType() == typeof(int) && mapping.IdType == typeof(long))
-                    return;
-
                 throw new InvalidOperationException(
-                    $"The id type for {typeof(T).FullName} is {mapping.IdType.Name}, but got {id.GetType().Name}");
+                    "The id type for {typeof(T).FullName} is {mapping.IdType.Name}, but got {id.GetType().Name}");
```
Comment (mysticmind) [inline src/Marten/QuerySession.cs:125]: Why was the `$` removed, isn't it not required for string interpolation?

File: src/Marten/QuerySession.cs [hunk 2]
```diff
@@ -119,13 +119,10 @@ private T load<T>(object id)
         private void assertCorrectIdType<T>(object id)
         {
             var mapping = Tenant.MappingFor(typeof(T));
-            if (id.GetType() != mapping.IdType)
+            if (id.GetType() != mapping.IdType && (id.GetType() != typeof(int) || mapping.IdType != typeof(long)))
```
Comment (mysticmind) [inline src/Marten/QuerySession.cs:122]: Adding this ` && (id.GetType() != typeof(int) || mapping.IdType != typeof(long)` in the parent `if` block looks to be changing the behaviour of earlier implementation. As per my understanding of earlier logic, if `id.GetType()` and `mapping.IdType` did not match, the subsequent check inside `if (id.GetType() == typeof(int) && mapping.IdType == typeof(long))` means that `int` and `long` are still compatible hence exception is not required to be thrown.

Can you please let me know your thoughts on the above?

File: src/Marten/QuerySession.cs (hunk 1 above)
Comment (oskardudycz) [reply src/Marten/QuerySession.cs:125]: You have good eyes! Nice catch. Fixed.

File: src/Marten/QuerySession.cs (hunk 2 above)
Comment (oskardudycz) [reply src/Marten/QuerySession.cs:122]: @mysticmind I just simplified the previous condition.

So I at first inverted the condition from line https://github.com/JasperFx/marten/pull/1428/files#diff-2f5f5939e30dd61deec7ee3492372bc9L124.

Then it was:
```
if (id.GetType() != mapping.IdType)
{
    if (id.GetType() != typeof(int) || mapping.IdType != typeof(long))
        throw new InvalidOperationException(
            $"The id type for {typeof(T).FullName} is {mapping.IdType.Name}, but got {id.GetType().Name}");
}
```
having that I joined two conditions to the final form

Comment (mysticmind) [reply src/Marten/QuerySession.cs:122]: The code in master is as below:
https://github.com/JasperFx/marten/blob/master/src/Marten/QuerySession.cs#L119|
```
        private void assertCorrectIdType<T>(object id)
        {
            var mapping = Tenant.MappingFor(typeof(T));
            if (id.GetType() != mapping.IdType)
            {
                if (id.GetType() == typeof(int) && mapping.IdType == typeof(long))
                    return;

                throw new InvalidOperationException(
                    $"The id type for {typeof(T).FullName} is {mapping.IdType.Name}, but got {id.GetType().Name}");
            }
        }
```

Comment (oskardudycz) [reply src/Marten/QuerySession.cs]: Reverted the change - I agree my version was too confusing,

Comment (mysticmind) [reply src/Marten/QuerySession.cs]: @oskardudycz  with ref to our DM. I was bit sleep deprived. The changes result in equivalent logic but the earlier logic was bit explicit on why we are running those checks. Thanks for reverting it.

**Pull request merged**